pkg/resource/stack/deployment.go

```diff
@@ -276,10 +276,24 @@ func DeserializeDeploymentV3(deployment apitype.DeploymentV3, secretsProv Secret
 		enc = e
 	}
 
+	sc := SecretsCache{}
```

Comment:
Do we need a new type here? Feels like we could do this without such a large refactoring by moving the caching into the Decrypter. We do something similar with the [`cachingCrypter`](https://github.com/pulumi/pulumi/blob/master/pkg/resource/stack/secrets.go#L68-L152) to avoid re-encrypting secrets during serialization.